### PR TITLE
Overhaul of GPIO module & adjusted Serial module accordingly.

### PR DESCRIPTION
**Pull request by arjanmels**
- Renamed GPIO and RTC_IO register field to strip common prefix.
- Fixed bugs in GPIO
- Simplified GPIO code

TODO:
- Add support for routing
- Add support for additional configuration (e.g. pullup/down, drive strength, etc.)

This depends on https://github.com/esp-rs/esp32/pull/34 and it builds on #34.

Addresses issue #35 and #19

### CODE REVIEW

**Comment by arjanmels** [comment]
Changed this to WIP so I can include more changes in one go.

**Comment by arjanmels** [comment]
@MabezDev and @chocol4te I think this is in pretty good shape now. Your review and feedback is most welcome. (Todo's are intterrupt handling and perhaps some more functionality on the RTC_MUX.)

(I also updated the i2c to teh latest version: https://github.com/arjanmels/esp32-hal/tree/feature-i2c-am

File: src/clock_control/mod.rs [hunk 1]
```diff
@@ -1459,4 +1267,87 @@ impl ClockControl {
             CPUSource::APLL => unimplemented!(),
         }
     }
+
+    /// Get RTC tick count since boot
+    ///
+    /// This function can usually take up to one RTC clock cycles (~300us).
```
Comment (fmckeogh) [inline src/clock_control/mod.rs]:
```suggestion
    /// This function can usually take up to one RTC clock cycle (~300us).
```

File: src/gpio/mod.rs [hunk 2]
```diff
@@ -0,0 +1,886 @@
...
+use {
+    crate::target::{GPIO, IO_MUX, RTCIO},
+    core::{convert::Infallible, marker::PhantomData},
+    embedded_hal::digital::v2::{OutputPin as _, StatefulOutputPin as _},
+};
+
+mod mux;
+pub use mux::*;
+
+/// Extension trait to split a GPIO peripheral in independent pins and registers
```
Comment (fmckeogh) [inline src/gpio/mod.rs]:
```suggestion
/// Extension trait to split a GPIO peripheral into independent pins and registers
```

File: src/gpio/mod.rs [hunk 3]
```diff
@@ -0,0 +1,886 @@
...
+pub trait GpioExt {
+    /// The to split the GPIO into
```
Comment (fmckeogh) [inline src/gpio/mod.rs]:
```suggestion
    /// The type to split the GPIO into
```

File: src/gpio/mod.rs [hunk 4]
```diff
@@ -0,0 +1,886 @@
...
+    /// Enable/Disable input circuitry while in sleep mode
+    fn enable_input_in_sleep_mode(&mut self, on: bool) -> &mut Self;
+
+    /// Connect input to peripheral using default options
+    ///
+    /// This is a wrapper around [connect_input_to_peripheral_with_options][
+    /// InputPin::connect_input_to_peripheral_with_options], which sets
+    /// all the options to false.
+
+    fn connect_input_to_peripheral(&mut self, signal: InputSignal) -> &mut Self;
```
Comment (fmckeogh) [inline src/gpio/mod.rs]:
```suggestion
    /// all the options to false.
    fn connect_input_to_peripheral(&mut self, signal: InputSignal) -> &mut Self;
```

File: src/gpio/mod.rs [hunk 5]
```diff
@@ -0,0 +1,886 @@
...
+                    AlternateFunction::Function3
+                }
+                else {
+                    match signal {
+                        $( $(
+                            OutputSignal::$af_signal => AlternateFunction::$af,
+                        )* )?
+                        _ => AlternateFunction::Function3
+                    }
+                };
```
Comment (fmckeogh) [inline src/gpio/mod.rs]:
```suggestion
                let af = if force_via_gpio_mux {
                    AlternateFunction::Function3
                } else {
                    match signal {
                        $( $(
                            OutputSignal::$af_signal => AlternateFunction::$af,
                        )* )?
                        _ => AlternateFunction::Function3
                    }
                };
```

File: src/gpio/mod.rs [hunk 6]
```diff
@@ -0,0 +1,886 @@
...
+
+                if af== AlternateFunction::Function3 && signal as usize > 256 {
```
Comment (fmckeogh) [inline src/gpio/mod.rs]:
```suggestion
                if af == AlternateFunction::Function3 && signal as usize > 256 {
```

File: src/gpio/mod.rs [hunk 7]
```diff
@@ -0,0 +1,886 @@
...
+                    panic!("Cannot connect this peripheral to GPIO");
+                }
+
+                self.set_alternate_function(af);
+
+                let clipped_signal= if (signal as usize) <= 256 {signal as u16} else {256};
```
Comment (fmckeogh) [inline src/gpio/mod.rs]:
```suggestion
                let clipped_signal = if signal as usize <= 256 { signal as u16 } else { 256 };
```

**Review by fmckeogh — CHANGES_REQUESTED** [review]
Just some really small comment/formatting changes suggested, looks really solid and can't think of anything to change in the implementation!

**Comment by arjanmels** [comment]
> Just some really small comment/formatting changes suggested, looks really solid and can't think of anything to change in the implementation!

Thanks, always nice to put the dots on the i's and cross the t's. :-)

**Comment by arjanmels** [comment]
In principle, I think this is ready for merge: https://github.com/esp-rs/esp32/pull/34 has been merged as well as #34. Also the problems @chocol4te was having are solved (https://github.com/esp-rs/esp32-hal/pull/26#issuecomment-648833861).

However, I could still add interrupt support and more extensive support for the RTC IO mux. I will be working on this over the next day/weekend; I can merge it in this PR or can make a new one.

**Comment by fmckeogh** [comment]
@arjanmels imho it might be slightly easier to merge this and have separate PRs for the additional functionality, but I’m also happy to just rebase against this branch when I finish up the I2C work :)

File: Cargo.toml [hunk 8]
```diff
@@ -32,12 +32,11 @@ rt = ["esp32/rt", "xtensa-lx6-rt"]
 [dependencies]
 esp32-hal-proc-macros = { path = "procmacros" }
 
-xtensa-lx6-rt = { version = "0.2.0", optional = true }
-xtensa-lx6 = "0.1.0"
-esp32 = { version = "0.5.0", default-features = false }
+xtensa-lx6-rt = { version = "0.3.0", optional = true }
+xtensa-lx6 = "0.2.0"
+esp32 = "0.5.0"
```
Comment (MabezDev) [inline Cargo.toml]:
```suggestion
esp32 = { version = "0.6.0", default-features = false }
```

**Review by MabezDev — CHANGES_REQUESTED** [review]
Great work @arjanmels ! I think it's best to merge this as is, as if the diffs get much bigger I'll need 1 hour to read through ;), but in all seriousness this is a well needed change! I also believe this approach is novel among Rust HAL designs so it will be interesting to see how it plays out!

**Pull request merged**